auth.go

```diff
@@ -0,0 +1,62 @@
+package sdk
+
+import (
+	"net/http"
+	"sync"
+)
+
+// BasicAuth basic authentication for the the OpenFaaS client
+type BasicAuth struct {
+	Username string
+	Password string
+}
+
+// Set Authorization Basic header on request
+func (auth *BasicAuth) Set(req *http.Request) error {
+	req.SetBasicAuth(auth.Username, auth.Password)
+	return nil
+}
+
+// A TokenSource is anything that can return an OIDC ID token that can be exchanged for
+// an OpenFaaS token.
+type TokenSource interface {
+	// Token returns a token or an error.
+	Token() (string, error)
+}
+
+// TokenAuth bearer token authentication for OpenFaaS deployments with OpenFaaS IAM
+// enabled.
+type TokenAuth struct {
+	// TokenURL represents the OpenFaaS gateways token endpoint URL.
+	TokenURL string
+
+	// TokenSource used to get an ID token that can be exchanged for an OpenFaaS ID token.
+	TokenSource TokenSource
+
+	mu    sync.Mutex // guards token
```

Comment:
We call this lock everywhere else, could we be consistent?

Reply:
Renamed this to lock.